src/storage-controller/src/lib.rs

```diff
@@ -1707,397 +1711,135 @@ where
-            Some((_id, m)) = clients.next() => m,
-        };
-
-        self.stashed_response = Some(msg);
-    }
-
-    #[instrument(level = "debug")]
-    async fn process(
+    #[instrument(level = "debug")]
+    async fn process(
         &mut self,
         storage_metadata: &StorageMetadata,
     ) -> Result<Option<Response<T>>, anyhow::Error> {
         let mut updated_frontiers = None;
+        let mut dropped_read_holds = Vec::new();
         match self.stashed_response.take() {
             None => (),
             Some(StorageResponse::FrontierUppers(updates)) => {
                 self.update_write_frontiers(&updates);
                 updated_frontiers = Some(Response::FrontierUpdates(updates));
             }
             Some(StorageResponse::DroppedIds(ids)) => {
-                let shards_to_finalize = ids
-                    .iter()
-                    .filter_map(|id| {
-                        // Note: All handles to the id should be dropped by now and the since of
-                        // the collection should be downgraded to the empty antichain. If handles
-                        // to the shard still exist, then we will incorrectly report the shard as
-                        // alive, and if the since of the shard has not been downgraded, then we
-                        // will continuously fail to finalize it.
-                        //
-                        // TODO(parkmycar): Should we be asserting that .remove(...) is some? In
-                        // other words that we know about the collection we're receiving an event
-                        // for.
-                        self.collections
-                            .remove(id)
-                            .map(|state| state.collection_metadata.data_shard)
-                    })
-                    .collect();
+                for id in ids.iter() {
+                    if let Some(collection) = self.collections.remove(id) {
+                        match collection.data_source {
+                            DataSource::Ingestion(_ingestion) => {
+                                let dropped_subsources = self
+                                    .dropped_ingestions
+                                    .remove(id)
+                                    .expect("missing dropped subsources");
+
+                                // The cluster is not sending these, so we take
+                                // matters into our own hands!
+                                tracing::debug!(?dropped_subsources, "synthesizing DroppedIds messages for subsources and the remap shard");
+                                self.internal_response_sender
+                                    .send(StorageResponse::DroppedIds(
+                                        dropped_subsources.into_iter().collect(),
+                                    ))
+                                    .expect("we are still alive");
```

Comment:
I don't see the code this is replacing...is this fixing a separate known issue as part of this pr? Additionally, why don't we just find the set of subsources/remap shard at this moment, instead of using this intermediate state? My understanding is that `self.collections` still has all that data at this point?

Reply:
It's new, and it fixes part of the issue that we're never cleaning up subsource state! I fixed this, because I was annoyed that the `StorageController` never drops the hold that it has with `StorageCollections`, meaning the latter never cleans up it's state.

> Additionally, why don't we just find the set of subsources/remap shard at this moment, instead of using this intermediate state? My understanding is that self.collections still has all that data at this point?

Unfortunately, we don't have that state anymore. `drop_sources_unvalidated` removes the `IngestionExport` from the `Ingestion`, and I didn't want to touch that flow more than necessary. 🙈 In `drop_sources_unvalidated`, I now basically move the state over, so that we still have it here where we then synthesize messages. That whole flow around `DroppedIds` needs some love (before this PR and certainly also after!).

Reply:
I agree, lets not touch flows more than necessary!

It's very clarifying to my reading of these changes to know this is new!

Reply:
I consider `DroppedIds` to be a separate part of the protocol. Petros (not summoning him here 😅) has opinions about how the protocol should evolve, so I don't want to put anything down here right now.